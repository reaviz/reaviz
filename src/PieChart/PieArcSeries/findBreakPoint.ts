type Coordinate = [number, number];

/**
 * Finds intermediate point between two points so that this three points
 * can be nicely connected by two lines. One of this lines must be horizontal
 */
export function findBreakPoint(
  [startX, startY]: Coordinate,
  [endX, endY]: Coordinate
): Coordinate {
  let breakPoint: Coordinate = [0, 0];

  // whether we should create breakpoint near pie or near label
  const breakPointCondition = (endY - startY) * Math.sign(startY) >= 0;

  if (breakPointCondition) {
    // extend the line starting from startY till the endY
    let scale = Math.abs(endY / startY) || 1;
    const minScale = 1;
    const maxScale = Math.abs(endX / startX) || 1;

    scale = Math.max(Math.min(maxScale, scale), minScale);

    breakPoint = [startX * scale, endY];
  } else {
    // some arbitrary scale to ensure that break point will be placed
    // at some horizontal distance from the end point
    let scale = 0.85;
    const minScale = Math.abs(startX / endX) || 1;
    const maxScale = 1;

    scale = Math.max(Math.min(maxScale, scale), minScale);

    breakPoint = [endX * scale, startY];
  }

  return breakPoint;
}
