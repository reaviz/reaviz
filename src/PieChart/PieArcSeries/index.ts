export * from './PieArc';
export * from './PieArcLabel';
export * from './PieArcSeries';
