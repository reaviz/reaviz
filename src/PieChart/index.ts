export * from './PieArcSeries';
export * from './PieChart';
