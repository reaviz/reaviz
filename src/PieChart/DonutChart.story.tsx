import React, { useState, Fragment } from 'react';
import { PieChart } from './PieChart';
import { categoryData, randomNumber } from 'reaviz-data-utils';
import { PieArc, PieArcLabel, PieArcSeries } from './PieArcSeries';
import { Gradient } from '@/common/Gradient';

export default {
  tags: ['snapshot'],
  title: 'Charts/Pie Chart/Donut',
  component: PieChart,
  subcomponents: {
    PieArc,
    PieArcLabel,
    PieArcSeries
  }
};

export const Simple = () => (
  <PieChart
    id="simple"
    width={350}
    height={250}
    data={categoryData}
    series={<PieArcSeries doughnut={true} colorScheme="cybertron" />}
  />
);

export const WithGradient = () => (
  <PieChart
    id="with-gradient"
    width={350}
    height={250}
    data={categoryData}
    series={
      <PieArcSeries
        doughnut={true}
        colorScheme="cybertron"
        arc={<PieArc gradient={<Gradient />} />}
      />
    }
  />
);

export const RoundedAndSpaced = () => (
  <PieChart
    id="rounded-and-spaced"
    width={350}
    height={250}
    data={categoryData}
    series={
      <PieArcSeries
        cornerRadius={4}
        padAngle={0.02}
        padRadius={200}
        doughnut={true}
        colorScheme="cybertron"
      />
    }
  />
);

RoundedAndSpaced.story = {
  name: 'Rounded and spaced'
};

export const Labels = () => (
  <PieChart
    id="labels"
    width={350}
    height={250}
    data={categoryData}
    series={<PieArcSeries doughnut={true} />}
  />
);

export const InnerLabel = () => (
  <div
    style={{
      position: 'relative',
      height: '250px',
      width: '350px',
      alignItems: 'center',
      display: 'flex',
      justifyContent: 'center'
    }}
  >
    <div style={{ position: 'absolute', top: 0, left: 0 }}>
      <PieChart
        id="inner-label"
        width={350}
        height={250}
        data={categoryData}
        series={
          <PieArcSeries doughnut={true} label={null} colorScheme="cybertron" />
        }
      />
    </div>
    <h2 style={{ margin: '0 5px', padding: 0, color: 'white' }}>
      {categoryData.length} Attacks
    </h2>
  </div>
);

const LiveUpdatingStory = () => {
  const [data, setData] = useState([...categoryData]);

  const updateData = () => {
    const newData = [...data];
    const updateCount = randomNumber(1, 4);

    let idx = 0;
    while (idx <= updateCount) {
      const updateIndex = randomNumber(0, data.length - 1);
      newData[updateIndex] = {
        ...newData[updateIndex],
        data: randomNumber(10, 100)
      };

      idx++;
    }

    setData(newData);
  };

  return (
    <Fragment>
      <PieChart width={350} height={250} data={data} />
      <br />
      <button onClick={updateData}>Update</button>
    </Fragment>
  );
};

const ArcLabel = React.memo(function ArcLabel({
  title,
  description,
  data,
  icon,
  textAnchor,
  showText
}: {
  title: string;
  description: string;
  data: number; // the chart data value we labeling
  icon: React.ReactElement | null;
  textAnchor: 'start' | 'end';
  showText: boolean;
}) {
  const iconContainer = (
    <div
      style={{
        [textAnchor === 'start' ? 'marginRight' : 'marginLeft']: showText
          ? '4px'
          : 0,
        width: '24px',
        height: '24px',
        fill: '#fff',
        flexShrink: 0
      }}
    >
      {icon}
    </div>
  );

  const ellipsis: React.CSSProperties = {
    whiteSpace: 'nowrap',
    textOverflow: 'ellipsis',
    overflow: 'hidden'
  };

  return (
    <div
      style={{
        margin: '0 4px',
        display: 'flex',
        alignItems: 'center',
        justifyContent: `flex-${textAnchor}`,
        lineHeight: 1,
        height: '24px',
        textAlign: textAnchor === 'start' ? 'left' : 'right'
      }}
    >
      {textAnchor === 'start' && iconContainer}
      <div style={{ minWidth: 0 }}>
        <div style={ellipsis}>
          {title} - {data}
        </div>
        <div style={ellipsis}>{description}</div>
      </div>
      {textAnchor === 'end' && iconContainer}
    </div>
  );
});
