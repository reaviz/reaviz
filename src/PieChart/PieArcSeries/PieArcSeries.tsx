import React, { FC, Fragment, ReactElement, useMemo } from 'react';
import { CloneElement } from 'reablocks';
import { ArcData } from '@/PieChart';
import { PieArc, PieArcProps } from './PieArc';
import {
  PIE_ARC_LABEL_DEFAULT_PROPS,
  PieArcLabel,
  PieArcLabelProps
} from './PieArcLabel';
import { getColor, ColorSchemeType } from '@/common/color';
import {
  calculateCentroid,
  calculateInnerArc,
  calculateLabelPositions,
  calculateRadius
} from './radiusUtils';
import { identifier } from 'safe-identifier';
import { mergeDefaultProps } from '@/common';

export interface PieArcSeriesProps {
  /**
   * Unique id for the series.
   */
  id?: string;

  /**
   * Animated set by the `PieArc` components.
   */
  animated: boolean;

  /**
   * Outer radius set by the parent component.
   */
  outerRadius: number;

  /**
   * Inner radius set by the parent component.
   */
  innerRadius: number;

  /**
   * Pad Angle between adjacent arcs, see https://github.com/d3/d3-shape#arc_padAngle
   */
  padAngle: number;

  /**
   * Pad Radius between adjacent arcs, see https://github.com/d3/d3-shape#arc_padRadius
   */
  padRadius: number;

  /**
   * Corner Radius of the arcs, see https://github.com/d3/d3-shape#arc_cornerRadius
   */
  cornerRadius: number;

  /**
   * Data set by the parent component.
   */
  data: ArcData[];

  /**
   * Width of the arc
   */
  arcWidth: number;

  /**
   * Doughnut, render as a donut shape
   */
  doughnut: boolean;

  /**
   * Explode: OuterRadius will be adjusted by the data property
   */
  explode: boolean;

  /**
   * Display all labels shows labels even if there is little space
   */
  displayAllLabels: boolean;

  /**
   * Label component
   */
  label?: ReactElement<PieArcLabelProps, typeof PieArcLabel> | null;

  /**
   * Arc Component
   */
  arc: ReactElement<PieArcProps, typeof PieArc>;

  /**
   * Color scheme
   */
  colorScheme: ColorSchemeType;

  /**
   * Height set by the parent component
   */
  height: number;

  /**
   * Width set by the parent component
   */
  width: number;
}

export const PieArcSeries: FC<Partial<PieArcSeriesProps>> = (props) => {
  const {
    doughnut,
    arcWidth,
    label,
    colorScheme,
    width,
    displayAllLabels,
    height,
    explode,
    id,
    animated,
    cornerRadius,
    padAngle,
    padRadius,
    arc,
    data
  } = mergeDefaultProps(PIE_ARC_SERIES_DEFAULT_PROPS, props);

  const labelProps = useMemo(
    () => ({ ...PIE_ARC_LABEL_DEFAULT_PROPS, ...label?.props }),
    [label]
  );

  const { outerRadius, innerRadius, labelWidth } = calculateRadius(
    height,
    width,
    label,
    arcWidth,
    doughnut
  );

  const innerArc = calculateInnerArc(
    data,
    innerRadius,
    outerRadius,
    cornerRadius,
    padAngle,
    padRadius,
    explode
  );

  const positions = label
    ? calculateLabelPositions(
        data,
        outerRadius,
        // 4 is for vertical margins between labels
        labelProps.height + 4,
        cornerRadius,
        padAngle,
        padRadius,
        displayAllLabels
      )
    : [];

  const centroid = calculateCentroid(data, innerRadius, outerRadius, explode);

  function renderItem(arcData: ArcData, index: number) {
    const safeKey = identifier(arcData.data.key.toString());
    const color = getColor({
      data,
      colorScheme,
      point: arcData.data,
      index
    });

    return (
      <Fragment key={safeKey}>
        {positions[index] && (
          <CloneElement<PieArcLabelProps>
            element={label}
            data={arcData}
            centroid={centroid}
            outerRadius={outerRadius}
            width={labelWidth}
            position={positions[index]}
          />
        )}
        <CloneElement<PieArcProps>
          id={`${id}-arc-${safeKey}`}
          data={arcData}
          animated={animated}
          arc={innerArc}
          color={color}
          element={arc}
        />
      </Fragment>
    );
  }

  return <>{data.map(renderItem)}</>;
};

export const PIE_ARC_SERIES_DEFAULT_PROPS = {
  animated: true,
  colorScheme: 'cybertron',
  innerRadius: 0,
  cornerRadius: 0,
  padAngle: 0,
  padRadius: 0,
  explode: false,
  displayAllLabels: false,
  arcWidth: 0.25,
  label: <PieArcLabel />,
  arc: <PieArc />
};
