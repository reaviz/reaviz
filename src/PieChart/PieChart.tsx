import React, { FC, ReactElement, useCallback, useMemo } from 'react';
import classNames from 'classnames';
import { PieArcDatum } from 'd3-shape';
import { pie } from 'd3-shape';
import { CloneElement, useId } from 'reablocks';
import {
  ChartProps,
  ChartContainer,
  ChartContainerChildProps
} from '@/common/containers';
import { ChartShallowDataShape } from '@/common/data';
import { PieArcSeries, PieArcSeriesProps } from './PieArcSeries';

export type ArcData = PieArcDatum<ChartShallowDataShape>;

export interface PieChartProps extends ChartProps {
  /**
   * Data the chart will receive to render.
   */
  data?: ChartShallowDataShape[];

  /**
   * Whether the chart is disabled.
   */
  disabled?: boolean;

  /**
   * Whether to display labels even if their value has a small display radius.
   */
  displayAllLabels?: boolean;

  /**
   * The series component that renders the arc components.
   */
  series?: ReactElement<PieArcSeriesProps, typeof PieArcSeries>;
}

export const PieChart: FC<PieChartProps> = ({
  id,
  width,
  height,
  className,
  containerClassName,
  displayAllLabels,
  data = [],
  margins = 10,
  series = <PieArcSeries />
}) => {
  const newId = useId(id);

  const internalData = useMemo(() => {
    const pieLayout = pie<void, ChartShallowDataShape>().value(
      (d: ChartShallowDataShape) => Number(d.data)
    );

    // Explode sort doesn't work right...
    if (!series.props.explode) {
      pieLayout.sort(null);
    }

    return pieLayout(data);
  }, [data, series]);

  const renderSeries = useCallback(
    ({ chartWidth, chartHeight }: ChartContainerChildProps) => {
      return (
        <CloneElement<PieArcSeriesProps>
          element={series}
          id={newId}
          data={internalData}
          height={chartHeight}
          width={chartWidth}
          displayAllLabels={displayAllLabels}
        />
      );
    },
    [displayAllLabels, internalData, newId, series]
  );

  return (
    <ChartContainer
      id={id}
      width={width}
      height={height}
      margins={margins}
      containerClassName={containerClassName}
      xAxisVisible={false}
      yAxisVisible={false}
      center={true}
      className={classNames(className)}
    >
      {renderSeries}
    </ChartContainer>
  );
};
