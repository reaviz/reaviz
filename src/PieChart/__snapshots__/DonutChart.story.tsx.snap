// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Charts/Pie Chart/Donut InnerLabel smoke-test 1`] = `
<div style="position: relative; height: 250px; width: 350px; align-items: center; display: flex; justify-content: center;">
  <div style="position: absolute; top: 0px; left: 0px;">
    <div class="_container_1u3dt_1"
         width="350"
         height="250"
         style="height: 250px; width: 350px;"
    >
      <svg width="350"
           height="250"
           class="_svg_1u3dt_30"
           tabindex="0"
      >
        <g transform="translate(175, 125)">
          <g tabindex="0"
             aria-label="DLP: 10"
             role="graphics-document"
          >
            <path role="graphics-symbol"
                  d="M0,-115A115,115,0,0,1,67.595,-93.037L50.696,-69.778A86.25,86.25,0,0,0,0,-86.25Z"
                  fill="#2d60e8"
                  style="cursor: initial;"
            >
            </path>
          </g>
          <g tabindex="0"
             aria-label="Malware: 30"
             role="graphics-document"
          >
            <path role="graphics-symbol"
                  d="M67.595,-93.037A115,115,0,0,1,67.595,93.037L50.696,69.778A86.25,86.25,0,0,0,50.696,-69.778Z"
                  fill="#2c77e0"
                  style="cursor: initial;"
            >
            </path>
          </g>
          <g tabindex="0"
             aria-label="IDS: 20"
             role="graphics-document"
          >
            <path role="graphics-symbol"
                  d="M67.595,93.037A115,115,0,0,1,-67.595,93.037L-50.696,69.778A86.25,86.25,0,0,0,50.696,69.778Z"
                  fill="#2b8dd8"
                  style="cursor: initial;"
            >
            </path>
          </g>
          <g tabindex="0"
             aria-label="Phishing Attack: 40"
             role="graphics-document"
          >
            <path role="graphics-symbol"
                  d="M-67.595,93.037A115,115,0,0,1,0,-115L0,-86.25A86.25,86.25,0,0,0,-50.696,69.778Z"
                  fill="#2aa1d1"
                  style="cursor: initial;"
            >
            </path>
          </g>
        </g>
      </svg>
    </div>
  </div>
  <h2 style="margin: 0px 5px; padding: 0px; color: white;">
    4 Attacks
  </h2>
</div>
`;

exports[`Charts/Pie Chart/Donut Labels smoke-test 1`] = `
<div class="_container_1u3dt_1"
     width="350"
     height="250"
     style="height: 250px; width: 350px;"
>
  <svg width="350"
       height="250"
       class="_svg_1u3dt_30"
       tabindex="0"
  >
    <g transform="translate(175, 125)">
      <g opacity="1">
        <title>
          DLP
        </title>
        <text dy=".35em"
              fill="#8F979F"
              font-size="11"
              font-family="sans-serif"
              text-anchor="start"
              style="shape-rendering: crispedges; transform: translate3d(92px, -87.4972px, 0px);"
        >
          DLP
        </text>
        <polyline fill="none"
                  stroke="rgba(127,127,127,0.5)"
                  points="20.729890039319393,-63.80004130146656,24.927370879579097,-76.7185589811424,28.429563482495162,-87.49719949915412,92,-87.49719949915412"
        >
        </polyline>
      </g>
      <g tabindex="0"
         aria-label="DLP: 10"
         role="graphics-document"
      >
        <path role="graphics-symbol"
              d="M0,-76.667A76.667,76.667,0,0,1,45.064,-62.025L33.798,-46.518A57.5,57.5,0,0,0,0,-57.5Z"
              fill="#2d60e8"
              style="cursor: initial;"
        >
        </path>
      </g>
      <g opacity="1">
        <title>
          Malware
        </title>
        <text dy=".35em"
              fill="#8F979F"
              font-size="11"
              font-family="sans-serif"
              text-anchor="start"
              style="shape-rendering: crispedges; transform: translate3d(92px, 0px, 0px);"
        >
          Malware
        </text>
        <polyline fill="none"
                  stroke="rgba(127,127,127,0.5)"
                  points="67.08333333333334,0,80.66666666666667,0,80.66666666666667,0,92,0"
        >
        </polyline>
      </g>
      <g tabindex="0"
         aria-label="Malware: 30"
         role="graphics-document"
      >
        <path role="graphics-symbol"
              d="M45.064,-62.025A76.667,76.667,0,0,1,45.064,62.025L33.798,46.518A57.5,57.5,0,0,0,33.798,-46.518Z"
              fill="#2c77e0"
              style="cursor: initial;"
        >
        </path>
      </g>
      <g opacity="1">
        <title>
          IDS
        </title>
        <text dy=".35em"
              fill="#8F979F"
              font-size="11"
              font-family="sans-serif"
              text-anchor="end"
              style="shape-rendering: crispedges; transform: translate3d(-92px, 92px, 0px);"
        >
          IDS
        </text>
        <polyline fill="none"
                  stroke="rgba(127,127,127,0.5)"
                  points="4.107669472140081e-15,67.08333333333334,4.9394087565609914e-15,80.66666666666667,5.633375276077825e-15,92,-92,92"
        >
        </polyline>
      </g>
      <g tabindex="0"
         aria-label="IDS: 20"
         role="graphics-document"
      >
        <path role="graphics-symbol"
              d="M45.064,62.025A76.667,76.667,0,0,1,-45.064,62.025L-33.798,46.518A57.5,57.5,0,0,0,33.798,46.518Z"
              fill="#2b8dd8"
              style="cursor: initial;"
        >
        </path>
      </g>
      <g opacity="1">
        <title>
          Phishing Attack
        </title>
        <text dy=".35em"
              fill="#8F979F"
              font-size="11"
              font-family="sans-serif"
              text-anchor="end"
              style="shape-rendering: crispedges; transform: translate3d(-92px, -28.4296px, 0px);"
        >
          Phishing Attack
        </text>
        <polyline fill="none"
                  stroke="rgba(127,127,127,0.5)"
                  points="-63.800041301466564,-20.729890039319383,-76.7185589811424,-24.927370879579083,-87.49719949915414,-28.42956348249515,-92,-28.42956348249515"
        >
        </polyline>
      </g>
      <g tabindex="0"
         aria-label="Phishing Attack: 40"
         role="graphics-document"
      >
        <path role="graphics-symbol"
              d="M-45.064,62.025A76.667,76.667,0,0,1,0,-76.667L0,-57.5A57.5,57.5,0,0,0,-33.798,46.518Z"
              fill="#2aa1d1"
              style="cursor: initial;"
        >
        </path>
      </g>
    </g>
  </svg>
</div>
`;

exports[`Charts/Pie Chart/Donut RoundedAndSpaced smoke-test 1`] = `
<div class="_container_1u3dt_1"
     width="350"
     height="250"
     style="height: 250px; width: 350px;"
>
  <svg width="350"
       height="250"
       class="_svg_1u3dt_30"
       tabindex="0"
  >
    <g transform="translate(175, 125)">
      <g opacity="1">
        <title>
          DLP
        </title>
        <text dy=".35em"
              fill="#8F979F"
              font-size="11"
              font-family="sans-serif"
              text-anchor="start"
              style="shape-rendering: crispedges; transform: translate3d(92px, -87.4972px, 0px);"
        >
          DLP
        </text>
        <polyline fill="none"
                  stroke="rgba(127,127,127,0.5)"
                  points="20.729890039319393,-63.80004130146656,24.927370879579097,-76.7185589811424,28.429563482495162,-87.49719949915412,92,-87.49719949915412"
        >
        </polyline>
      </g>
      <g tabindex="0"
         aria-label="DLP: 10"
         role="graphics-document"
      >
        <path role="graphics-symbol"
              d="M2,-72.419A4,4,0,0,1,6.33,-76.405A76.667,76.667,0,0,1,39.788,-65.534A4,4,0,0,1,40.949,-59.763L34.358,-50.693A4,4,0,0,1,29.098,-49.594A57.5,57.5,0,0,0,5.61,-57.226A4,4,0,0,1,2,-61.207Z"
              fill="#2d60e8"
              style="cursor: initial;"
        >
        </path>
      </g>
      <g opacity="1">
        <title>
          Malware
        </title>
        <text dy=".35em"
              fill="#8F979F"
              font-size="11"
              font-family="sans-serif"
              text-anchor="start"
              style="shape-rendering: crispedges; transform: translate3d(92px, 0px, 0px);"
        >
          Malware
        </text>
        <polyline fill="none"
                  stroke="rgba(127,127,127,0.5)"
                  points="67.08333333333334,0,80.66666666666667,0,80.66666666666667,0,92,0"
        >
        </polyline>
      </g>
      <g tabindex="0"
         aria-label="Malware: 30"
         role="graphics-document"
      >
        <path role="graphics-symbol"
              d="M44.185,-57.412A4,4,0,0,1,50.031,-58.092A76.667,76.667,0,0,1,50.031,58.092A4,4,0,0,1,44.185,57.412L37.594,48.342A4,4,0,0,1,38.175,42.999A57.5,57.5,0,0,0,38.175,-42.999A4,4,0,0,1,37.594,-48.342Z"
              fill="#2c77e0"
              style="cursor: initial;"
        >
        </path>
      </g>
      <g opacity="1">
        <title>
          IDS
        </title>
        <text dy=".35em"
              fill="#8F979F"
              font-size="11"
              font-family="sans-serif"
              text-anchor="end"
              style="shape-rendering: crispedges; transform: translate3d(-92px, 92px, 0px);"
        >
          IDS
        </text>
        <polyline fill="none"
                  stroke="rgba(127,127,127,0.5)"
                  points="4.107669472140081e-15,67.08333333333334,4.9394087565609914e-15,80.66666666666667,5.633375276077825e-15,92,-92,92"
        >
        </polyline>
      </g>
      <g tabindex="0"
         aria-label="IDS: 20"
         role="graphics-document"
      >
        <path role="graphics-symbol"
              d="M40.949,59.763A4,4,0,0,1,39.788,65.534A76.667,76.667,0,0,1,-39.788,65.534A4,4,0,0,1,-40.949,59.763L-34.358,50.693A4,4,0,0,1,-29.098,49.594A57.5,57.5,0,0,0,29.098,49.594A4,4,0,0,1,34.358,50.693Z"
              fill="#2b8dd8"
              style="cursor: initial;"
        >
        </path>
      </g>
      <g opacity="1">
        <title>
          Phishing Attack
        </title>
        <text dy=".35em"
              fill="#8F979F"
              font-size="11"
              font-family="sans-serif"
              text-anchor="end"
              style="shape-rendering: crispedges; transform: translate3d(-92px, -28.4296px, 0px);"
        >
          Phishing Attack
        </text>
        <polyline fill="none"
                  stroke="rgba(127,127,127,0.5)"
                  points="-63.800041301466564,-20.729890039319383,-76.7185589811424,-24.927370879579083,-87.49719949915414,-28.42956348249515,-92,-28.42956348249515"
        >
        </polyline>
      </g>
      <g tabindex="0"
         aria-label="Phishing Attack: 40"
         role="graphics-document"
      >
        <path role="graphics-symbol"
              d="M-44.185,57.412A4,4,0,0,1,-50.031,58.092A76.667,76.667,0,0,1,-6.33,-76.405A4,4,0,0,1,-2,-72.419L-2,-61.207A4,4,0,0,1,-5.61,-57.226A57.5,57.5,0,0,0,-38.175,42.999A4,4,0,0,1,-37.594,48.342Z"
              fill="#2aa1d1"
              style="cursor: initial;"
        >
        </path>
      </g>
    </g>
  </svg>
</div>
`;

exports[`Charts/Pie Chart/Donut Simple smoke-test 1`] = `
<div class="_container_1u3dt_1"
     width="350"
     height="250"
     style="height: 250px; width: 350px;"
>
  <svg width="350"
       height="250"
       class="_svg_1u3dt_30"
       tabindex="0"
  >
    <g transform="translate(175, 125)">
      <g opacity="1">
        <title>
          DLP
        </title>
        <text dy=".35em"
              fill="#8F979F"
              font-size="11"
              font-family="sans-serif"
              text-anchor="start"
              style="shape-rendering: crispedges; transform: translate3d(92px, -87.4972px, 0px);"
        >
          DLP
        </text>
        <polyline fill="none"
                  stroke="rgba(127,127,127,0.5)"
                  points="20.729890039319393,-63.80004130146656,24.927370879579097,-76.7185589811424,28.429563482495162,-87.49719949915412,92,-87.49719949915412"
        >
        </polyline>
      </g>
      <g tabindex="0"
         aria-label="DLP: 10"
         role="graphics-document"
      >
        <path role="graphics-symbol"
              d="M0,-76.667A76.667,76.667,0,0,1,45.064,-62.025L33.798,-46.518A57.5,57.5,0,0,0,0,-57.5Z"
              fill="#2d60e8"
              style="cursor: initial;"
        >
        </path>
      </g>
      <g opacity="1">
        <title>
          Malware
        </title>
        <text dy=".35em"
              fill="#8F979F"
              font-size="11"
              font-family="sans-serif"
              text-anchor="start"
              style="shape-rendering: crispedges; transform: translate3d(92px, 0px, 0px);"
        >
          Malware
        </text>
        <polyline fill="none"
                  stroke="rgba(127,127,127,0.5)"
                  points="67.08333333333334,0,80.66666666666667,0,80.66666666666667,0,92,0"
        >
        </polyline>
      </g>
      <g tabindex="0"
         aria-label="Malware: 30"
         role="graphics-document"
      >
        <path role="graphics-symbol"
              d="M45.064,-62.025A76.667,76.667,0,0,1,45.064,62.025L33.798,46.518A57.5,57.5,0,0,0,33.798,-46.518Z"
              fill="#2c77e0"
              style="cursor: initial;"
        >
        </path>
      </g>
      <g opacity="1">
        <title>
          IDS
        </title>
        <text dy=".35em"
              fill="#8F979F"
              font-size="11"
              font-family="sans-serif"
              text-anchor="end"
              style="shape-rendering: crispedges; transform: translate3d(-92px, 92px, 0px);"
        >
          IDS
        </text>
        <polyline fill="none"
                  stroke="rgba(127,127,127,0.5)"
                  points="4.107669472140081e-15,67.08333333333334,4.9394087565609914e-15,80.66666666666667,5.633375276077825e-15,92,-92,92"
        >
        </polyline>
      </g>
      <g tabindex="0"
         aria-label="IDS: 20"
         role="graphics-document"
      >
        <path role="graphics-symbol"
              d="M45.064,62.025A76.667,76.667,0,0,1,-45.064,62.025L-33.798,46.518A57.5,57.5,0,0,0,33.798,46.518Z"
              fill="#2b8dd8"
              style="cursor: initial;"
        >
        </path>
      </g>
      <g opacity="1">
        <title>
          Phishing Attack
        </title>
        <text dy=".35em"
              fill="#8F979F"
              font-size="11"
              font-family="sans-serif"
              text-anchor="end"
              style="shape-rendering: crispedges; transform: translate3d(-92px, -28.4296px, 0px);"
        >
          Phishing Attack
        </text>
        <polyline fill="none"
                  stroke="rgba(127,127,127,0.5)"
                  points="-63.800041301466564,-20.729890039319383,-76.7185589811424,-24.927370879579083,-87.49719949915414,-28.42956348249515,-92,-28.42956348249515"
        >
        </polyline>
      </g>
      <g tabindex="0"
         aria-label="Phishing Attack: 40"
         role="graphics-document"
      >
        <path role="graphics-symbol"
              d="M-45.064,62.025A76.667,76.667,0,0,1,0,-76.667L0,-57.5A57.5,57.5,0,0,0,-33.798,46.518Z"
              fill="#2aa1d1"
              style="cursor: initial;"
        >
        </path>
      </g>
    </g>
  </svg>
</div>
`;

exports[`Charts/Pie Chart/Donut WithGradient smoke-test 1`] = `
<div class="_container_1u3dt_1"
     width="350"
     height="250"
     style="height: 250px; width: 350px;"
>
  <svg width="350"
       height="250"
       class="_svg_1u3dt_30"
       tabindex="0"
  >
    <g transform="translate(175, 125)">
      <g opacity="1">
        <title>
          DLP
        </title>
        <text dy=".35em"
              fill="#8F979F"
              font-size="11"
              font-family="sans-serif"
              text-anchor="start"
              style="shape-rendering: crispedges; transform: translate3d(92px, -87.4972px, 0px);"
        >
          DLP
        </text>
        <polyline fill="none"
                  stroke="rgba(127,127,127,0.5)"
                  points="20.729890039319393,-63.80004130146656,24.927370879579097,-76.7185589811424,28.429563482495162,-87.49719949915412,92,-87.49719949915412"
        >
        </polyline>
      </g>
      <g tabindex="0"
         aria-label="DLP: 10"
         role="graphics-document"
      >
        <path role="graphics-symbol"
              d="M0,-76.667A76.667,76.667,0,0,1,45.064,-62.025L33.798,-46.518A57.5,57.5,0,0,0,0,-57.5Z"
              fill="url(#gradient-with-gradient-arc-DLP)"
              style="cursor: initial;"
        >
        </path>
        <linearGradient spreadmethod="pad"
                        id="gradient-with-gradient-arc-DLP"
                        y1="0%"
                        y2="0%"
                        x1="0%"
                        x2="100%"
        >
          <stop offset="0%"
                stop-opacity="0.3"
                stop-color="#2d60e8"
          >
          </stop>
          <stop offset="80%"
                stop-opacity="1"
                stop-color="#2d60e8"
          >
          </stop>
        </linearGradient>
      </g>
      <g opacity="1">
        <title>
          Malware
        </title>
        <text dy=".35em"
              fill="#8F979F"
              font-size="11"
              font-family="sans-serif"
              text-anchor="start"
              style="shape-rendering: crispedges; transform: translate3d(92px, 0px, 0px);"
        >
          Malware
        </text>
        <polyline fill="none"
                  stroke="rgba(127,127,127,0.5)"
                  points="67.08333333333334,0,80.66666666666667,0,80.66666666666667,0,92,0"
        >
        </polyline>
      </g>
      <g tabindex="0"
         aria-label="Malware: 30"
         role="graphics-document"
      >
        <path role="graphics-symbol"
              d="M45.064,-62.025A76.667,76.667,0,0,1,45.064,62.025L33.798,46.518A57.5,57.5,0,0,0,33.798,-46.518Z"
              fill="url(#gradient-with-gradient-arc-Malware)"
              style="cursor: initial;"
        >
        </path>
        <linearGradient spreadmethod="pad"
                        id="gradient-with-gradient-arc-Malware"
                        y1="0%"
                        y2="0%"
                        x1="0%"
                        x2="100%"
        >
          <stop offset="0%"
                stop-opacity="0.3"
                stop-color="#2c77e0"
          >
          </stop>
          <stop offset="80%"
                stop-opacity="1"
                stop-color="#2c77e0"
          >
          </stop>
        </linearGradient>
      </g>
      <g opacity="1">
        <title>
          IDS
        </title>
        <text dy=".35em"
              fill="#8F979F"
              font-size="11"
              font-family="sans-serif"
              text-anchor="end"
              style="shape-rendering: crispedges; transform: translate3d(-92px, 92px, 0px);"
        >
          IDS
        </text>
        <polyline fill="none"
                  stroke="rgba(127,127,127,0.5)"
                  points="4.107669472140081e-15,67.08333333333334,4.9394087565609914e-15,80.66666666666667,5.633375276077825e-15,92,-92,92"
        >
        </polyline>
      </g>
      <g tabindex="0"
         aria-label="IDS: 20"
         role="graphics-document"
      >
        <path role="graphics-symbol"
              d="M45.064,62.025A76.667,76.667,0,0,1,-45.064,62.025L-33.798,46.518A57.5,57.5,0,0,0,33.798,46.518Z"
              fill="url(#gradient-with-gradient-arc-IDS)"
              style="cursor: initial;"
        >
        </path>
        <linearGradient spreadmethod="pad"
                        id="gradient-with-gradient-arc-IDS"
                        y1="0%"
                        y2="0%"
                        x1="0%"
                        x2="100%"
        >
          <stop offset="0%"
                stop-opacity="0.3"
                stop-color="#2b8dd8"
          >
          </stop>
          <stop offset="80%"
                stop-opacity="1"
                stop-color="#2b8dd8"
          >
          </stop>
        </linearGradient>
      </g>
      <g opacity="1">
        <title>
          Phishing Attack
        </title>
        <text dy=".35em"
              fill="#8F979F"
              font-size="11"
              font-family="sans-serif"
              text-anchor="end"
              style="shape-rendering: crispedges; transform: translate3d(-92px, -28.4296px, 0px);"
        >
          Phishing Attack
        </text>
        <polyline fill="none"
                  stroke="rgba(127,127,127,0.5)"
                  points="-63.800041301466564,-20.729890039319383,-76.7185589811424,-24.927370879579083,-87.49719949915414,-28.42956348249515,-92,-28.42956348249515"
        >
        </polyline>
      </g>
      <g tabindex="0"
         aria-label="Phishing Attack: 40"
         role="graphics-document"
      >
        <path role="graphics-symbol"
              d="M-45.064,62.025A76.667,76.667,0,0,1,0,-76.667L0,-57.5A57.5,57.5,0,0,0,-33.798,46.518Z"
              fill="url(#gradient-with-gradient-arc-Phishing_Attack)"
              style="cursor: initial;"
        >
        </path>
        <linearGradient spreadmethod="pad"
                        id="gradient-with-gradient-arc-Phishing_Attack"
                        y1="0%"
                        y2="0%"
                        x1="0%"
                        x2="100%"
        >
          <stop offset="0%"
                stop-opacity="0.3"
                stop-color="#2aa1d1"
          >
          </stop>
          <stop offset="80%"
                stop-opacity="1"
                stop-color="#2aa1d1"
          >
          </stop>
        </linearGradient>
      </g>
    </g>
  </svg>
</div>
`;
